Move coreclr EH second pass to native code

janvorli opened the pull request:
There were some GC holes discovered caused by the fact that GC can be triggered during 2nd pass of EH in-between calls to finally handlers and catch handler. After considering options, moving the 2nd pass to native code seems the most reasonable solution.

janvorli committed:
[WIP] Move coreclr EH second pass to native code

There were some GC holes discovered caused by the fact that GC can be
triggered during 2nd pass of EH in-between calls to finally handlers and
catch handler. After considering options, moving the 2nd pass to native
code seems the most reasonable solution.

davidwrighton commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -4174,6 +4206,179 @@ Exit:;
+
+    // Search the clauses for one that contains the current offset.
+    RhEHClause ehClause;
+    for (uint curIdx = 0; EHEnumNextWorker(&ehEnum, &ehClause, /* onlyFinallys */ true) && curIdx < idxLimit; curIdx++)
```
I don't believe the `onlyFinallys` logic is actually a good way to implement this. `idxLimit` is based on the index Including the catch clauses, but this will search a different amount. I would instead suggest that the parameter be something like `doNotCalculateCatchType`, and just have the function fill in the `_pTargetType` will NULL if its set to true.

janvorli commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -3121,18 +3135,21 @@ PropagateForeignExceptionThroughNativeFrames(IN     PEXCEPTION_RECORD   pExcepti
+    CONTRACTL
+    {
+        MODE_COOPERATIVE;
+        GC_NOTRIGGER;
```
It has always been that way, but you are right it is not correct. The same for the fact that it has been NOTHROW

janvorli commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -3530,8 +3542,8 @@ extern "C" CLR_BOOL QCALLTYPE EHEnumNext(EH_CLAUSE_ENUMERATOR* pEHEnum, RhEHClau
         else if (flags & COR_ILEXCEPTION_CLAUSE_FAULT)
         {
-            EH_LOG((LL_INFO100, " fault clause\n"));
             pEHClause->_clauseKind = RH_EH_CLAUSE_FAULT;
+            EH_LOG((LL_INFO100, " fault clause\n"));
```
It doesn't have any negative effect. This was the only EH_LOG that was not at an end of a block here, so I've taken the opportunity to unify it.

janvorli commented:
And one more thing not done yet is rethrow.

janvorli committed:
Several fixes

* Reflect PR feedback
* Implement rethrow
* Implement new way of collided unwind detection now that the
  CallCatchFunclet is not called via pinvoke
* Remove forced reporting of EH code from stack frame iterator, as we
  now cannot have that code on the stack during 2nd pass

janvorli commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs:
```diff
@@ -869,6 +897,12 @@ private static void DispatchEx(scoped ref StackFrameIterator frameIter, ref ExIn
+        }
+
+        [StackTraceHidden]
+        private static void DispatchExPass2(ref ExInfo exInfo, UIntPtr handlingFrameSP, byte* pCatchHandler, IntPtr pReversePInvokePropagationCallback, IntPtr pReversePInvokePropagationContext)
```
Renamed

janvorli committed:
Fix arm64 build and remove now useless stuff from stackwalk

janvorli committed:
Fix build break

jkotas commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs:
```diff
@@ -881,15 +940,15 @@ private static void DispatchEx(scoped ref StackFrameIterator frameIter, ref ExIn
             // 'main body' without first encountering the funclet.  The thunks used to invoke 2nd-pass
             // funclets will always toggle this mode off before invoking them.
-#if NATIVEAOT
             InternalCalls.RhpSetThreadDoNotTriggerGC();
```
```suggestion
            InternalCalls.RhpSetThreadDoNotTriggerGC();

```

janvorli committed:
Fix offsets, exception interception and MUSL build break

janvorli committed:
Alternative change with minimalistic managed code differences

janvorli committed:
Fix 32 bit offsets

davidwrighton commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -3466,16 +3396,12 @@ extern "C" CLR_BOOL QCALLTYPE EHEnumInitFromStackFrameIterator(StackFrameIterato
 }
 
-extern "C" CLR_BOOL QCALLTYPE EHEnumNext(EH_CLAUSE_ENUMERATOR* pEHEnum, RhEHClause* pEHClause)
+// The onlyFinallys option makes the function return only finally and fault clauses. It is used in
```
comment out of date

davidwrighton commented:
I would add some correlation comments so that we know that C++ function X is supposed to be the same as C# function Y. But otherwise, this actually looks good to me (assuming it works of course)

janvorli committed:
Fix arm64 issue

jkotas committed:
Update src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs

janvorli committed:
Add/fix some contracts

janvorli commented:
@jkotas I've added some contracts. I also had to change contract on the `Debugger::FirstChanceManagedExceptionCatcherFound` from `GC_TRIGGERS_FROM_GETJITINFO` to `GC_NOTRIGGER`. I cannot find how the GC can get triggered by the GetJitInfo (and it has GC_NOTRIGGER contract), so I hope I haven't missed anything.

janvorli commented:
There is one issue left that I am investigating - Linux arm64 failures in several libraries tests.

janvorli committed:
Fix another arm64 issue - need to use adjusted PC

kg commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -3121,18 +3063,19 @@ PropagateForeignExceptionThroughNativeFrames(IN     PEXCEPTION_RECORD   pExcepti
+    CONTRACTL
+    {
+        MODE_COOPERATIVE;
+        GC_NOTRIGGER;
```
If it's not too much trouble, what's the reasoning behind this function being GC_NOTRIGGER and then having GC contract violations in it? Compatibility with existing code?

janvorli replied:
The point was that the whole function should not trigger GC, but the actual catch handler is managed code, so obviously GC can happen in it. So the contract violation helped me to express that.
However, I am changing that in my upcoming commit after hitting some issue with profiler and thinking again about when the GC is prohibited. Looking at the legacy EH code, the GC is prohibited there only in-between finallys in the same frame. That makes sense and I am moving to that model. Without it, profiler notifications would be a problem, as each frame we visit sends a profiler notification and that switches to preemptive GC mode, thus GC can happen there. So we cannot prohibit GC for the whole 2nd pass.

janvorli committed:
Change the GC prevention to single frame only

janvorli committed:
Fix one new contract issue

janvorli committed:
Update GC forbid locations

This is done to ensure that no GC is allowed between the scanned stack
range is extended and a funclet for the current frame is called.

janvorli commented:
@jkotas, @davidwrighton the current state of this PR is final from my point of view. Can you please review it?

kg commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -3466,16 +3401,12 @@ extern "C" CLR_BOOL QCALLTYPE EHEnumInitFromStackFrameIterator(StackFrameIterato
 
-extern "C" CLR_BOOL QCALLTYPE EHEnumNext(EH_CLAUSE_ENUMERATOR* pEHEnum, RhEHClause* pEHClause)
+// The doNotCalculateCatchType option makes the function skip calculation of the catch type. It is used in
+// the 2nd path of EH to avoid possible GC stemming from a call to ResolveEHClause.
```
nit: path -> pass?

janvorli commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -3364,33 +3301,31 @@ extern "C" void QCALLTYPE ResumeAtInterceptionLocation(REGDISPLAY* pvRegDisplay)
     MethodDesc *pMD = exInfo->m_frameIter.m_crawl.GetFunction();
     // Profiler, debugger and ETW events
     TADDR spForDebugger = GetSpForDiagnosticReporting(pvRegDisplay);
     exInfo->MakeCallbacksRelatedToHandler(true, pThread, pMD, &exInfo->m_CurrentClause, (DWORD_PTR)pHandlerIP, spForDebugger);
```
The callbacks that are called here don't trigger GC, there is a subset of profiler callbacks that are documented here https://github.com/dotnet/runtime/blob/fc1dee66fac8ae088d01d07b4a2faaeef15a234a/docs/design/coreclr/botr/profiling.md#gc-safe-callouts
that runtime calls without switching to preemptive mode.

davidwrighton reviewed (approved):
I believe this is correct and will fix the gc issue. I wish this invariant could be better documented/tested.

janvorli committed:
Fix/add comment based on review

jkotas commented on src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.cs:
```diff
@@ -557,6 +557,9 @@ internal object ThrownException
             internal volatile UIntPtr _handlingFrameSP;
 
 #if TARGET_ARM64
+            // On ARM64, two frames can have the same SP, when a leaf function
```
Does this actually happen in our current codegen? I thought that we always emit non-empty frames currently:

```
            stp     fp, lr, [sp, #-0x10]!
            mov     fp, sp
            ldp     fp, lr, [sp], #0x10
            ret     lr
```

If this is a future proofing in case we stop doing this, I think this is needed on all current architectures except x86/x64.

janvorli replied:
This was there since May 2018. This PR doesn't change it in any way. Here is the PR that brought that in: https://github.com/dotnet/corert/pull/5560. So it is possible that it is not needed for NativeAOT / CoreCLR and it is just a remainder of handling the behavior of the pre-nativeaot compiler.
Let me run coreclr / libraries tests on arm64 device with this removed to see if there are any issues.
